src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -471,3 +469,65 @@ lemma ideal.homogeneous_hull_eq_Inf (I : ideal A) :
 eq.symm $ is_glb.Inf_eq $ (ideal.homogeneous_hull.gc 𝒜).is_least_l.is_glb
 
 end galois_connection
+
+section irrelevant_ideal
+
+open graded_algebra set_like.graded_monoid direct_sum
+
+variables [comm_semiring R] [semiring A]
+variables [algebra R A] [decidable_eq ι]
+variables [canonically_ordered_add_monoid ι]
+variables (𝒜 : ι → submodule R A) [graded_algebra 𝒜]
+
+/--
+If `A` is graded by a canonically ordered add monoid, then the projection map `x ↦ x₀` is a ring
+homomorphism.
+-/
+def graded_algebra.proj_zero_ring_hom : A →+* A :=
+{ to_fun := λ a, decompose 𝒜 a 0,
+  map_one' := decompose_of_mem_same 𝒜 one_mem,
+  map_zero' := by simp only [subtype.ext_iff_val, map_zero, zero_apply, submodule.coe_zero],
+  map_add' := λ _ _, by simp [subtype.ext_iff_val, map_add, add_apply, submodule.coe_add],
+  map_mul' := λ x y,
+    have m : ∀ x, x ∈ supr 𝒜, from λ x, (is_internal 𝒜).supr_eq_top.symm ▸ submodule.mem_top,
+    begin
+    refine submodule.supr_induction 𝒜 (m x) (λ i c hc, _) _ _,
+    { refine submodule.supr_induction 𝒜 (m y) (λ j c' hc', _) _ _,
+      { by_cases h : i + j = 0,
+        { rw [decompose_of_mem_same 𝒜 (show c * c' ∈ 𝒜 0, from h ▸ mul_mem hc hc'),
+            decompose_of_mem_same 𝒜 (show c ∈ 𝒜 0, from (add_eq_zero_iff.mp h).1 ▸ hc),
+            decompose_of_mem_same 𝒜 (show c' ∈ 𝒜 0, from (add_eq_zero_iff.mp h).2 ▸ hc')] },
+        { rw [decompose_of_mem_ne 𝒜 (mul_mem hc hc') h],
+          cases (show i ≠ 0 ∨ j ≠ 0, by rwa [add_eq_zero_iff, not_and_distrib] at h) with h' h',
+          { simp only [decompose_of_mem_ne 𝒜 hc h', zero_mul] },
+          { simp only [decompose_of_mem_ne 𝒜 hc' h', mul_zero] } } },
+      { simp only [map_zero, zero_apply, submodule.coe_zero, mul_zero] },
+      { intros _ _ hd he, simp only [mul_add, map_add, add_apply, submodule.coe_add, hd, he] } },
+    { simp only [map_zero, zero_apply, submodule.coe_zero, zero_mul] },
+    { rintros _ _ ha hb, simp only [add_mul, map_add, add_apply, submodule.coe_add, ha, hb] },
+  end }
+
+/--
+For a graded ring `⨁ᵢ 𝒜ᵢ` graded by a `canonically_ordered_add_monoid ι`, the irrelevant ideal
+refers to `⨁_{i≥0} 𝒜ᵢ`, or equivalently `{a | a₀ = 0}`.
+
+This definition is used in Proj construction where `ι` is always `ℕ` so the irrelevant ideal is
+simply elements with `0` as 0-th coordinate. But the notion of irrelevant ideal makes sense in a
+more general setting by defining it as the ideal of elements with `0` as i-th coordinate for all
+`i ≤ 0`, i.e. `{a | ∀ (i : ι), i ≤ 0 → aᵢ = 0 }`.
```

Comment:
The definition that you give here is not the same right? Or am I missing something?
By defining it as the kernel of projection onto the `0`th component, you don't require that `aᵢ = 0` for all `i ≤ 0`.

Reply:
Here `ι` is assumed to be a `canonically_ordered_add_monoid`, so there is nothing below zero. I only intended to use this for `ι = natural numbers` in Proj construction. But the notion makes sense for `ι` with negative part by asking irrelevant ideal to be the ideal of elements with `0` nonpositive coordinates. Since I don't intend to use any indexing set, so I didn't write down the most general definition.

Reply:
May be I should write the docstring to explicitly mention this as potential future work